# AutoTransOP: translating omics signatures without orthologue requirements using deep learning
Github repository of the study:
> AutoTransOP: Translating Omics Signatures without Orthologue Requirements using Deep Learning <br>
> Meimetis, N., Pullen, K. M., Zhu, D. Y., Nilsson, A., Hoang, T. N., Magliacane, S., & Lauffenburger, D. A.</sup>

Published in NPJ Systems Biology and Applications: https://doi.org/10.1038/s41540-024-00341-9

This repository is administered by the Lauffenburger Lab and @NickMeim. For questions contact meimetis@mit.edu

**Trained models of this study are too big to be uploaded here and are available upon reasonable request.**

The development of effective therapeutics and vaccines for human diseases requires a systematic understanding of human biology. While animal and in vitro culture models have successfully elucidated the molecular mechanisms of diseases in many studies, they yet fail to adequately recapitulate human biology as evidenced by the predominant likelihood of failure in clinical trials. To address this broadly important problem, we developed AutoTransOP, a neural network autoencoder framework to map omics profiles from designated species or cellular contexts into a global latent space, from which germane information can be mapped between different contexts. This approach performs as well or better than extant machine learning methods and can identify animal/culture-specific molecular features predictive of other contexts, without requiring homology matching. For an especially challenging test case, we successfully apply our framework to a set of inter-species vaccine serology studies, where no 1-1 mapping between human and non-human primate features exists.

The current repository contains code for:
1. Initial evaluation of the quality and preprocessing of the data.
2. Embedding evaluations of the autoencoder models.
3. Deep learning models (autoencoder's approaches) to translate transcriptomic profiles between different cell types.
4. Importance estimation of each feature for specific tasks.
5. Code to re-create the results of the research article.

## Data
The transcriptomic signatures (level 5 profiles) of the L1000 CMap resource[^1] are used for this study, together with data from the Bioconductor resource[^2].

The transcriptomic profiles were generated by measuring 978 important (landmark) genes in cancer with a Luminex bead-based assay and computationally inferring the rest[^1]. 

**Details on how to access these data can be found in the data folder**, but generally the main resources can be accessed [here](https://clue.io/data/CMap2020?fbclid=IwAR1Uc379nDYELH8lYU9MPI9TiAT3054_55g72Ymbgm7FAW7WZnPD3YBCXeI#LINCS2020)

## Folder structure
1. **article_supplementary_info** : Folder containing code to re-create the supplementary figures and tables of the article
2. **data** : Folder that should contain the raw data of the study.
3. **figures** : Folder containing the scripts to produce the figures of the study (except for the serology and fibrosis case studies).
4. **fibrosis** : Folder containing code and data to re-create the results of the article regarding the fibrosis case study.
5. **learning** : Folder containing machine and deep learning algorithms and models and scripts to estimate genes' importance as considered by each model.
6. **preprocessing** : Folder containing scripts to pre-process the raw data and evaluate their quality.
	* **preprocessed_data** : Here the pre-processed data to be used in the subsequent analysis are stored.
7. **results** : Here the results of a subsequent analysis should be stored.
8. **postprocessing** : Folder containing scripts to evaluate models' embeddings and genes' importance.
9. **serology** : Folder containing code and data to re-create the results of the article regarding the serology case study.

## Installation
The study utilizes multiple resources from the Python and R programming languages.

**R dependencies**: 
You can check the list below and manually install your preferences.

**Important Note:**
* **This installation was performed in a WINDOWS environment.** 
* **For a Linux installation there might be needed some manual installation of external dependencies (especially) for tidyverse. Please check libraries' documentation online**
* **For a MAC installation we encourage checking online**

In a quick overview, the following R libraries and versions (**although any version of the following libraries is appropriate**) were/are used to produce the figures and results of the study:
1. [R](https://cran.r-project.org/bin/windows/base/) version 4.1.2
2. [tidyverse](https://www.tidyverse.org/packages/) 1.3.1
3. [BiocManager](https://www.bioconductor.org/install/) 1.30.16
4. [cmapR](https://bioconductor.org/packages/release/bioc/html/cmapR.html) 1.4.0
5. [org.Hs.eg.db](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html) 3.13.0
6. [rhdf5](https://bioconductor.org/packages/release/bioc/html/rhdf5.html) 2.36.0
7. [doFuture](https://cran.r-project.org/web/packages/doFuture/index.html) 0.12.0
8. [doRNG](https://cran.r-project.org/web/packages/doRNG/index.html) 1.8.2
9. [ggplot2](https://ggplot2.tidyverse.org/) 3.3.5
10. [ggpubr](https://www.rdocumentation.org/packages/ggpubr/versions/0.4.0) 0.4.0
11. [GeneExpressionSignature](https://www.bioconductor.org/packages/release/bioc/html/GeneExpressionSignature.html) 1.38.0
12. [caret](https://cran.r-project.org/web/packages/caret/index.html) 6.0-94
13. [Rtsne](https://cran.r-project.org/web/packages/Rtsne/index.html) 0.16
14. [factoextra](https://cran.r-project.org/web/packages/factoextra/index.html) 1.0.7
15. [ggpubr](https://rpkgs.datanovia.com/ggpubr/) 0.6.0
16. [ggpattern](https://coolbutuseless.github.io/package/ggpattern/) 1.1.0
17. [ggridges](https://cran.r-project.org/web/packages/ggridges/vignettes/introduction.html) 0.5.4
18. [ggrepel](https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html) 0.9.3
19. [rstatix](https://cran.r-project.org/web/packages/rstatix/index.html) 0.7.2
20. [patchwork](https://patchwork.data-imaginist.com/) 1.1.2.9000
21. [dorothea](https://saezlab.github.io/dorothea/) 1.4.2
22. [fgsea](https://bioconductor.org/packages/release/bioc/html/fgsea.html) 1.18.0
23. [AnnotationDbi](https://bioconductor.org/packages/release/bioc/html/AnnotationDbi.html) 1.54.1
24. [EGSEAdata](http://bioconductor.org/packages/release/data/experiment/html/EGSEAdata.html) 1.20.0
25. [topGO](http://bioconductor.org/packages/release/bioc/html/topGO.html) 2.44.0
26. [GO.db](https://bioconductor.org/packages/release/data/annotation/html/GO.db.html) 3.13.0

**Python dependencies**: 
First, install conda (anaconda) environment in your computer and then you can use the commands **in a bash-terminal** after the list of libraries.

**Important Note:**
* **Pytorch GPU installation CHANGES according to your NVIDIA GPU and cuda version. Check the [pytorch installation guide](https://pytorch.org/get-started/locally/) here for more information.**
* **This installation was performed in a WINDOWS environment. For other environments check libraries' documentation** 

In a quick overview, the following Python libraries and versions (**although different versions are POSSIBLY also appropriate**) were/are used:
1. [python](https://www.python.org/downloads/) 3.8.8
2. [seaborn](https://seaborn.pydata.org/installing.html) 0.11.2 (version does not matter for this library)
3. [numpy](https://numpy.org/install/) 1.20.3 (version does not matter for this library)
4. [pandas](https://pandas.pydata.org/docs/getting_started/install.html) 1.3.5 (version does not matter for this library)
5. [matplotlib](https://anaconda.org/conda-forge/matplotlib) 3.5.1 (version does not matter for this library)
6. [scipy](https://anaconda.org/anaconda/scipy) 1.7.3
7. [scikit-learn](https://scikit-learn.org/stable/install.html) 1.0.2
8. CUDA Version: 11.1 - 11.5 **(Very important for pytorch GPU installation)**
9. [pytorch](https://pytorch.org/get-started/locally/) **(Important to download a compatible version for your system. Click on pytorch to view more information on its original website)**
10. GPU: NVIDIA GeForce RTX 3060

```bash
# After installing anaconda create a conda environment:
conda create -n myenv python=3.8.8
conda install numpy
conda install pandas
conda install -c conda-forge scikit-learn
conda install -c conda-forge matplotlib
conda install seaborn
conda install -c anaconda scipy
# This is hardware-specific
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch
```


## References
[^1]: Subramanian, Aravind, et al. "A next generation connectivity map: L1000 platform and the first 1,000,000 profiles." Cell 171.6 (2017): 1437-1452.
[^2]: Gentleman, Robert C., et al. "Bioconductor: open software development for computational biology and bioinformatics." Genome biology 5.10 (2004): 1-16.
